learning_phase/chat_sample/client.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"strings"
+)
+
+//various commands numbered from 0, 1, 2, 3...
+const (
+	cmdauth = iota
+	cmdpm
+	cmdmsg
+	cmdquit
+)
+
+type command struct {
+	id     int
+	client *client
+	args   []string
+}
+
+type client struct {
+	conn     net.Conn
+	address  string
+	username string
+	commands chan<- command //incoming channel
+}
+
+func (c *client) readInput(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done(): //when cancel is called
```

Comment:
So, reading user input from terminal is a blocking operation - so once the select statement enters the deafult: section, it'll be stuck there until the user enters an input and the rest of the code in that section is executed. So, if cancel() is called on the context in a parent function, it won't be received until the client has entered something, the remaining processing in the loop is finished and the select statement is encountered again in the next iteration. Hope it makes sense.

So how do you think this can be fixed?

Reply:
Fixed this by having another channel for receiving client message. it will be called as go routine for the readInput function. So now that problem won't occur it will terminate as soon as cancel is received.